### PR TITLE
fuzz: add target for `ScriptPubKeyMan` (legacy)

### PR DESCRIPTION
**Pull request by brunoerg**
This PR adds target for `{Legacy}ScriptPubKeyMan`. I'm working on a descriptor one and will do it in a separate file. I tried to focus here on functions that we use directly and we may have in some unit tests.

### CODE REVIEW

File: src/wallet/test/fuzz/scriptpubkeyman.cpp [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    FuzzedDataProvider fuzzed_data_provider{buffer.data(), buffer.size()};
+    const auto& node{g_setup->m_node};
+    Chainstate* chainstate = &node.chainman->ActiveChainstate();
+    CWallet& wallet = *g_wallet_ptr;
+    {
+        LOCK(wallet.cs_wallet);
+        wallet.SetLastBlockProcessed(chainstate->m_chain.Height(), chainstate->m_chain.Tip()->GetBlockHash());
+    }
+
+    LegacyScriptPubKeyMan& legacy_spkm{*wallet.GetOrCreateLegacyScriptPubKeyMan()};
```
[the remark this answers is not in the document]
Comment (brunoerg) [reply src/wallet/test/fuzz/scriptpubkeyman.cpp:39]: Yea, I just realized it. I'm changing it!

File: src/wallet/test/fuzz/scriptpubkeyman.cpp [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    static const auto testing_setup = MakeNoLogFileContext<const TestingSetup>();
+    g_setup = testing_setup.get();
+}
+
+FUZZ_TARGET(scriptpubkeyman, .init = initialize_spkm)
+{
+    FuzzedDataProvider fuzzed_data_provider{buffer.data(), buffer.size()};
+    const auto& node{g_setup->m_node};
+    Chainstate* chainstate = &node.chainman->ActiveChainstate();
+    std::unique_ptr<CWallet> wallet{std::make_unique<CWallet>(node.chain.get(), "", CreateMockableWalletDatabase())};
```
[the remark this answers is not in the document]
Comment (brunoerg) [reply src/wallet/test/fuzz/scriptpubkeyman.cpp:30]: I think the contexts are different. `notifications` is working with descriptor and here is legacy, I believe I'll have something better when I finish the `DescriptorScriptPubKey` target. Also, I believe the `FuzzedWallet` in `notifications` is disabled for spkm.

**Comment by brunoerg** [comment]
CI failure seems unrelated.

**Comment by brunoerg** [comment]
Pushed to re-run CI.

**Review by Ayush170-Future — APPROVED** [review]
ACK

I support the idea of having separate files for Legacy and Descriptor because I think the Wallet implementation would differ. Overall, the code looks great to me.

File: src/wallet/test/fuzz/scriptpubkeyman.cpp [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    LegacyScriptPubKeyMan& legacy_spkm{*wallet->GetOrCreateLegacyScriptPubKeyMan()};
+    SignatureData data;
+    CScript watch_script;
+
+    LIMITED_WHILE(fuzzed_data_provider.ConsumeBool(), 10000) {
+        CallOneOf(
+            fuzzed_data_provider,
+            [&] {
+                CKey key;
+                key.MakeNewKey(fuzzed_data_provider.ConsumeBool());
```
[the remark this answers is not in the document]
Comment (brunoerg) [reply src/wallet/test/fuzz/scriptpubkeyman.cpp]: Yea, just changed the approach to use buffer.

File: src/wallet/test/fuzz/scriptpubkeyman.cpp [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+            [&] {
+                CScript script{ConsumeScript(fuzzed_data_provider)};
+                (void)legacy_spkm.IsMine(script);
+            },
+            [&] {
+                CScript script{ConsumeScript(fuzzed_data_provider)};
+                (void)legacy_spkm.CanProvide(script, data);
+            },
+            [&] {
+                CKeyID address(ConsumeUInt160(fuzzed_data_provider));
```
[the remark this answers is not in the document]
Comment (brunoerg) [reply src/wallet/test/fuzz/scriptpubkeyman.cpp]: Yes, just changed it.

**Review by maflcko — APPROVED** [review]
lgtm

File: src/wallet/test/fuzz/scriptpubkeyman.cpp [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+            [&] {
+                CScript script{ConsumeScript(fuzzed_data_provider)};
+                (void)legacy_spkm.IsMine(script);
+            },
+            [&] {
+                CScript script{ConsumeScript(fuzzed_data_provider)};
+                (void)legacy_spkm.CanProvide(script, data);
+            },
+            [&] {
+                key = keys[fuzzed_data_provider.ConsumeIntegralInRange<size_t>(0, num_keys - 1)];
```
[the remark this answers is not in the document]
Comment (brunoerg) [reply src/wallet/test/fuzz/scriptpubkeyman.cpp:69]: Yea, more elegant. Pushed changing it

**Comment by achow101** [comment]
Closing as there is little support for working on fixing non-critical issues in legacy wallets.

**Comment by brunoerg** [comment]
> Would be nice to have a descriptor wallet target? Alternatively, I think your plan was to speed up wallet_notifications IIRC?

Yes, I'm working on speeding up wallet_notifications.

**Pull request closed without merging**